runtime/common/src/check_nonce.rs

```diff
@@ -0,0 +1,285 @@
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, Dispatchable, One, SignedExtension},
+	transaction_validity::{
+		InvalidTransaction, TransactionLongevity, TransactionValidity, TransactionValidityError, ValidTransaction,
+	},
+};
+use sp_std::vec;
+
+/// Nonce check and increment to give replay protection for transactions.
+///
+/// # Transaction Validity
+///
+/// This extension affects `requires` and `provides` tags of validity, but DOES NOT
+/// set the `priority` field. Make sure that AT LEAST one of the signed extension sets
+/// some kind of priority upon validating transactions.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct CheckNonce<T: frame_system::Config + module_evm::Config>(
+	#[codec(compact)] pub T::Index,
+	#[codec(skip)] pub bool, /* should check evm nonce */
+);
+
+impl<T: frame_system::Config + module_evm::Config> CheckNonce<T> {
+	/// utility constructor. Used only in client/factory code.
+	pub fn from(nonce: T::Index) -> Self {
+		Self(nonce, false)
+	}
+}
+
+impl<T: frame_system::Config + module_evm::Config> sp_std::fmt::Debug for CheckNonce<T> {
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "CheckNonce({}, evm: {})", self.0, self.1)
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+impl<T: frame_system::Config + module_evm::Config> SignedExtension for CheckNonce<T>
+where
+	T::Call: Dispatchable<Info = DispatchInfo>,
+	T::AddressMapping: AddressMapping<T::AccountId>,
+{
+	type AccountId = T::AccountId;
+	type Call = T::Call;
+	type AdditionalSigned = ();
+	type Pre = ();
+	const IDENTIFIER: &'static str = "CheckNonce";
+
+	fn additional_signed(&self) -> sp_std::result::Result<(), TransactionValidityError> {
+		Ok(())
+	}
+
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		_call: &Self::Call,
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> Result<(), TransactionValidityError> {
+		let mut account = frame_system::Account::<T>::get(who);
+		if self.1 {
+			// should check evm nonce
+			let address = <T as module_evm::Config>::AddressMapping::get_evm_address(who)
+				.unwrap_or_else(|| <T as module_evm::Config>::AddressMapping::get_default_evm_address(who));
+			let account = module_evm::Accounts::<T>::get(&address)
+				.unwrap_or_else(|| module_evm::AccountInfo::<T::Index>::new(Default::default(), None));
+			if self.0 != account.nonce {
+				return Err(if self.0 < account.nonce {
+					InvalidTransaction::Stale
+				} else {
+					InvalidTransaction::Future
+				}
+				.into());
+			}
+		} else if self.0 != account.nonce {
+			return Err(if self.0 < account.nonce {
+				InvalidTransaction::Stale
+			} else {
+				InvalidTransaction::Future
+			}
+			.into());
+		}
+		account.nonce += T::Index::one();
+		frame_system::Account::<T>::insert(who, account);
```

Comment:
EvmNonce shouldn't do this.

Reply:
We must consider the increase of `evm.nonce`. We use `transactional` on `eth_call`, if the transaction fails, `evm.nonce` will not increase, then we will see the same nonce of 2 transactions.

Reply:
@zjb0807 yes I'm thinking about that. There's also the case when tx isn't evm call but it still needs to increment nonce